README.md
# aidockermon

[![Build Status](https://travis-ci.org/WqyJh/aidockermon.svg?branch=master)](https://travis-ci.org/WqyJh/aidockermon)
[![license](https://img.shields.io/badge/LICENCE-GPLv3-brightgreen.svg)](https://raw.githubusercontent.com/WqyJh/aidockermon/master/LICENSE)

Monitor system load of the server running the nvidia/cuda docker containers.

## Feature

- sysinfo: system static info
- sysload: system cpu/memory load
- gpu: nvidia gpu load
- disk: disk load
- containers: containers' load that based on the nvidia/cuda image

## Prerequisite

Python >= 3

## Installation

```bash
pip install aidockermon
```

Or use `setuptools`
```bash
python setup.py install
```

## Usage

```
$ aidockermon -h
usage: aidockermon [-h] [-v] {query,create-esindex,delete-esindex} ...

optional arguments:
  -h, --help            show this help message and exit
  -v, --version         show program's version number and exit

command:
  {query,create-esindex,delete-esindex}
    query               Query system info, log them via syslog protocol
    create-esindex      Create elasticsearch index
    delete-esindex      Delete elasticsearch index
```

```
$ aidockermon query -h
usage: aidockermon query [-h] [-l] [-r REPEAT] [-f FILTERS [FILTERS ...]] type

positional arguments:
  type                  info type: sysinfo, sysload, gpu, disk, containers

optional arguments:
  -h, --help            show this help message and exit
  -l, --stdout          Print pretty json to console instead of send a log
  -r REPEAT, --repeat REPEAT
                        n/i repeat n times every i seconds
  -f FILTERS [FILTERS ...], --filters FILTERS [FILTERS ...]
                        Filter the disk paths for disk type; filter the
                        container names for containers type
```

**For example:**

Show sysinfo

```bash
$ aidockermon query -l sysinfo
{
    "gpu": {
        "gpu_num": 2,
        "driver_version": "410.104",
        "cuda_version": "10.0"
    },
    "mem_tot": 67405533184,
    "kernel": "4.4.0-142-generic",
    "cpu_num": 12,
    "docker": {
        "version": "18.09.3"
    },
    "system": "Linux"
}
```

Show sys load

```bash
$ aidockermon query -l sysload
{
    "mem_free": 11866185728,
    "mem_used": 8023793664,
    "cpu_perc": 57.1,
    "mem_perc": 12.8,
    "mem_avail": 58803163136,
    "mem_tot": 67405533184
}
```

Show gpu load

```bash
$ aidockermon query -l gpu
{
    "mem_tot": 11177,
    "gpu_temperature": 76.0,
    "mem_free": 1047,
    "mem_used": 10130,
    "gpu_perc": 98.0,
    "gpu_id": 0,
    "mem_perc": 46.0
}
{
    "mem_tot": 11178,
    "gpu_temperature": 66.0,
    "mem_free": 3737,
    "mem_used": 7441,
    "gpu_perc": 95.0,
    "gpu_id": 1,
    "mem_perc": 44.0
}
```

Show disk usage

```bash
$ aidockermon query disk -l -f /
{
    "path": "/",
    "device": "/dev/nvme0n1p3",
    "total": 250702176256,
    "used": 21078355968,
    "free": 216865271808,
    "percent": 8.9
}

$ aidockermon query disk -l -f / /disk
{
    "path": "/",
    "device": "/dev/nvme0n1p3",
    "total": 250702176256,
    "used": 21078355968,
    "free": 216865271808,
    "percent": 8.9
}
{
    "path": "/disk",
    "device": "/dev/sda1",
    "total": 1968874311680,
    "used": 1551374692352,
    "free": 317462949888,
    "percent": 83.0
}
```

Show containers' load

Note that the `app_name` would be read from environment variable `APP_NAME`, which
is a short description for this training program.

```bash
$ aidockermon query containers -l -f DianAI
{
    "proc_name": "python3 test_run.py",
    "app_name": "测试程序",
    "pid": 13540,
    "container": "DianAI",
    "started_time": 1554698236,
    "running_time": 9343,
    "mem_used": 9757
}
{
    "proc_name": "python train.py",
    "app_name": "",
    "pid": 15721,
    "container": "DianAI",
    "started_time": 1554698236,
    "running_time": 19343,
    "mem_used": 1497
}
{
    "mem_limit": 67481047040,
    "net_output": 47863240948,
    "block_read": 1327175626752,
    "net_input": 18802869033,
    "mem_perc": 14.637655604461704,
    "block_write": 132278439936,
    "name": "DianAI",
    "cpu_perc": 0.0,
    "mem_used": 9877643264
}
```

## Config

### logging
```yaml
debug: false
log:
  version: 1

  # This is the default level, which could be ignored.
  # CRITICAL = 50
  # FATAL = CRITICAL
  # ERROR = 40
  # WARNING = 30
  # WARN = WARNING
  # INFO = 20
  # DEBUG = 10
  # NOTSET = 0
  #level: 20
  disable_existing_loggers: false
  formatters:
    simple:
      format: '%(levelname)s %(message)s'
    monitor:
      format: '%(message)s'
  filters:
    require_debug_true:
      (): 'aidockermon.handlers.RequireDebugTrue'
  handlers:
    console:
      level: DEBUG
      class: logging.StreamHandler
      formatter: simple
      filters: [require_debug_true]
    monitor:
      level: INFO
      class: rfc5424logging.handler.Rfc5424SysLogHandler
      address: [127.0.0.1, 1514]
      enterprise_id: 1
  loggers:
    runtime:
      handlers: [console]
      level: DEBUG
      propagate: false
    monitor:
      handlers: [monitor, console]
      level: INFO
      propagate: false

```

This is the default config, which should be located at `/etc/aidockermon/config.yml`.

You can modify the `address` value to specify the logging target.
- `address: [127.0.0.1, 1514]`: UDP to 127.0.0.1:1514
- `address: /var/log/aidockermon`: unix domain datagram socket

If you add an `socktype` argument, you can specify whether to use UDP or TCP as transport protocol.
- `socktype: 1`: TCP
- `socktype: 2`: UDP

Enable TLS/SSL:
```yaml
tls_enable: true
tls_verify: true
tls_ca_bundle: /path/to/ca-bundle.pem
```

Set `debug` as `true`, you can see message output in the console.

### Cronjob

```bash
sudo cp etc/cron.d/aidockermon /etc/cron.d
sudo systemctl restart cron
```

### syslog-ng

Using syslog-ng to collect logs and send them to elasticsearch 
for future use such as visualization with kibana.

```bash
cp etc/syslog-ng/syslog-ng.conf /etc/syslog-ng/
sudo systemctl restart syslog-ng
```

Sample config:

```bash
@version: 3.20

destination d_elastic {
	elasticsearch2(
		index("syslog-ng")
		type("${.SDATA.meta.type}")
		flush-limit("0")
		cluster("es-syslog-ng")
		cluster-url("http://localhost:9200")
		client-mode("http")
		client-lib-dir(/usr/share/elasticsearch/lib)
		template("${MESSAGE}\n")
	);
};

source s_python {
  #unix-dgram("/var/log/aidockermon");
	syslog(ip(127.0.0.1) port(1514) transport("udp") flags(no-parse));
};

log {
	source (s_python);
  parser { syslog-parser(flags(syslog-protocol)); };
	destination (d_elastic);
};
```
Modify it to specify the elasticsearch server and the log source's port and protocol.



CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [0.10.3](https://github.com/WqyJh/aidockermon/compare/v0.10.2...v0.10.3)

> 16 October 2019

- Fix type [`#4`](https://github.com/WqyJh/aidockermon/pull/4)
- Bump version to 0.10.3 [`b57822b`](https://github.com/WqyJh/aidockermon/commit/b57822b0c5202bc5ba5e42995d1749027f99573f)
- fix: fix type [`781588c`](https://github.com/WqyJh/aidockermon/commit/781588cde74cad745ee47e6fbed61252edb9316b)

#### 0.10.2

> 16 October 2019

- Release 0.10 [`#3`](https://github.com/WqyJh/aidockermon/pull/3)
- Deploy [`#2`](https://github.com/WqyJh/aidockermon/pull/2)
- feat(containers): add app_name [`#1`](https://github.com/WqyJh/aidockermon/pull/1)
- Update grafana.json [`b88157f`](https://github.com/WqyJh/aidockermon/commit/b88157f758898aab98848e69fe94a2c3fd6a941d)
- Initial commit [`3c526c2`](https://github.com/WqyJh/aidockermon/commit/3c526c2894fb0383d8db6671a176cc3460ad6766)
- Move mappings to etc/mappings.json [`7c9de3a`](https://github.com/WqyJh/aidockermon/commit/7c9de3abc1df493606f86a3f6b61bb9e71de7528)


test.sh
curl -X POST -H 'Content-Type: application/json' http://localhost:9200/syslog-ng-app/_mapping/app -d '
{
    "properties": {
        "app": {
            "properties": {
                "app_name": {
                    "type": "text",
                    "fielddata" : true,
                    "fields": {
                        "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                        }
                    }
                },
                "proc_name": {
                    "type": "text",
                    "fielddata" : true,
                    "fields": {
                        "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                        }
                    }
                },
                "pid": {
                    "type": "long"
                },
                "create_time": {
                    "type": "long"
                },
                "running_time": {
                    "type": "long"
                },
                "mem_used": {
                    "type": "long"
                },
                "timestamp": {
                    "type": "date"
                },
                "host": {
                    "type": "text",
                    "fielddata" : true,
                    "fields": {
                        "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                        }
                    }
                }
            }
        }
    }
}'


docker run --network host -v /tmp/syslog-ng.conf:/etc/syslog-ng/syslog-ng.conf balabit/syslog-ng -d

# qiming3
sudo /tmp/venv/bin/aidockermon query containers -l -f DianAI

aidockermon/utils.py
import json
import requests
import pkg_resources

ES_URL = 'http://localhost:9200'

ES_MAPPINGS = json.loads(pkg_resources.resource_string('etc', 'mappings.json').decode())

INDEX_PREFIX = 'syslog-ng-'


def es_res_ack(res):
    data = json.loads(res.content.decode())
    if data.get('acknowledged', False):
        return 'Succeed'
    else:
        return 'Failed'


def create_es_index(index, mappings=ES_MAPPINGS, es_url=ES_URL):
    type = index
    index = INDEX_PREFIX + type
    print('Creating index <%s>' % (index,), end=' ')
    req_url = '%s/%s' % (es_url, index)
    data = json.dumps({
        'mappings': {
            type: mappings[type],
        },
    })
    headers = {
        'Content-Type': 'application/json',
    }
    res = requests.put(req_url, data=data, headers=headers)

    print(es_res_ack(res))


def create_es_indicies(mappings=ES_MAPPINGS, es_url=ES_URL):
    for k in mappings.keys():
        create_es_index(k, es_url=es_url)


def delete_es_index(index, es_url=ES_URL):
    index = INDEX_PREFIX + index
    print('Deleting index <%s>' % (index,), end=' ')
    req_url = '%s/%s' % (es_url, index)
    res = requests.delete(req_url)

    print(es_res_ack(res))


def delete_es_indicies(mappings=ES_MAPPINGS, es_url=ES_URL):
    for k in mappings.keys():
        delete_es_index(k, es_url=es_url)


aidockermon/settings.py
import io
import yaml
import logging

DEBUG = False

LOGGING = {
    'version': 1,
    'level': logging.INFO,
    'disable_existing_loggers': False,
    'formatters': {
        # 'verbose': {
        #     'format': '%(levelname)s %(asctime)s %(module)s %(process)d %(thread)d %(message)s'
        # },
        'simple': {
            'format': '%(levelname)s %(message)s'
        },
        'monitor': {
            'format': '%(message)s'
        }
    },
    'filters': {
        'require_debug_true': {
            '()': 'aidockermon.handlers.RequireDebugTrue',
        },
    },
    'handlers': {
        'console': {
            'level': 'DEBUG',
            'class': 'logging.StreamHandler',
            'formatter': 'simple',
            'filters': ['require_debug_true'],
        },
        'monitor': {
            'level': 'INFO',
            'class': 'rfc5424logging.handler.Rfc5424SysLogHandler',
            # 'class': 'logging.handlers.SysLogHandler',

            # Use unix domain socket for better performance. (For production purpose)
            'address': '/var/log/aidockermon',

            # Use ip & port to enable packet sniff. (For debug purpose)
            # 'address': ('127.0.0.1', 1514),

            # `enterprise_id` MUST be set if you want to send structured_data.
            # Value of it would be ignored if you send data under
            # the predefined key like `meta`, which is unimportant
            # and could be any value in this case.
            'enterprise_id': 1,
        },
    },
    'loggers': {
        'runtime': {
            'handlers': ['console', ],
            'level': 'DEBUG',
            'propagate': False,
        },
        'monitor': {
            'handlers': ['monitor', 'console'],
            'level': 'INFO',
            'propagate': False,
        }
    },
}

CFG_FILE = '/etc/aidockermon/config.yml'


def load_yaml_compat(f):
    conf = yaml.load(f, Loader=yaml.SafeLoader)

    def _compat_log(log_conf):
        '''
        Because the Rfc5424SysLogHandler support tuple instead of
        list as it's address argument, we have to convert the
        address to tuple.
        '''
        if 'handlers' in log_conf:
            for k, v in log_conf['handlers'].items():
                if 'address' in v and isinstance(v['address'], list):
                    v['address'] = tuple(v['address'])

    if conf and 'log' in conf:
        _compat_log(conf['log'])

    return conf


try:
    with io.open(CFG_FILE, 'r') as f:
        conf = load_yaml_compat(f)

        if conf:
            DEBUG = conf.get('debug', False)
            LOGGING = conf.get('log', LOGGING)
except OSError as e:
    # print('Failed to parse config file: %s, use default config' % CFG_FILE)
    # Be quiet about it. No config file is a common scene.
    pass


aidockermon/aidockermon.py
#!/usr/bin/env python3

import os
import io
import sys
import csv
import json
import docker
import psutil
import logging
import platform
import argparse
import subprocess
import logging.config

from string import Template
from aidockermon import utils
from aidockermon import settings
from rfc5424logging import Rfc5424SysLogHandler

from aidockermon import __version__


logging.config.dictConfig(settings.LOGGING)
logger_runtime = logging.getLogger('runtime')
logger_monitor = logging.getLogger('monitor')


g_pre_stats = {}


class DeltaTemplate(Template):
    delimiter = '%'


def strfdelta(tdelta, fmt='%D %H:%M:%S'):
    d = {"D": tdelta.days}
    d['H'], rem = divmod(tdelta.seconds, 3600)
    d['M'], d['S'] = divmod(rem, 60)
    t = DeltaTemplate(fmt)
    return t.substitute(**d)


def strfcreate(tcreate, fmt='%D %H:%M:%S'):
    import datetime

    now = datetime.datetime.now()
    start = datetime.datetime.fromtimestamp(tcreate)
    return strfdelta(now - start)


def get_running_time(tcreate):
    import time

    tnow = time.time()
    return int(tnow - tcreate)


def fastfail_call(args):
    output = None
    try:
        output = subprocess.check_output(args, timeout=10).decode()
    except subprocess.CalledProcessError as e:
        logger_runtime.error(e)
    except OSError as e:
        logger_runtime.error(e)
    return output


def docker_info():
    client = docker.from_env()
    version = client.version()

    return {
        'version': version['Version'],  # client version
    }


def nvidia_info():
    '''
    nvidia-smi -q --display=COMPUTE

    ==============NVSMI LOG==============

    Timestamp                           : Sun Mar 31 09:55:41 2019
    Driver Version                      : 410.78
    CUDA Version                        : 10.0

    Attached GPUs                       : 2
    GPU 00000000:01:00.0
        Compute Mode                    : Default

    GPU 00000000:02:00.0
        Compute Mode                    : Default
    '''
    import re

    output = fastfail_call(['nvidia-smi', '-q', '--display=COMPUTE'])
    if output is None:
        return {}

    driver_version = re.search(
        'Driver Version[\s\S]+?(\d+\.\d+)', output).group(1)
    cuda_version = re.search('CUDA Version[\s\S]+?(\d+\.\d+)', output).group(1)
    gpu_num = int(re.search('Attached GPUs[\s\S]+?(\d+)', output).group(1))
    return {
        'driver_version': driver_version,
        'cuda_version': cuda_version,
        'gpu_num': gpu_num,
    }


def sys_static_info():
    import platform
    uname = platform.uname()

    return {
        'system': uname.system,
        'kernel': uname.release,
        'cpu_num': psutil.cpu_count(),
        'mem_tot': psutil.virtual_memory().total,
        'gpu': nvidia_info(),
        'docker': docker_info(),
    }


def container_pids(container_name):
    output = fastfail_call(['docker', 'top', container_name, '-eo', 'pid'])
    if output is None:
        return []

    return [int(x) for x in output.split()[1:]]


def nvidia_smi_query_gpu():
    '''
    nvidia-smi --query-gpu=utilization.gpu,utilization.memory,memory.used,memory.free,memory.total,temperature.gpu --format=csv,noheader,nounits
    0, 0, 0, 11178, 11178, 45
    100, 69, 9099, 2079, 11178, 63

    nvidia-smi --query-gpu=utilization.gpu,utilization.memory,memory.used,memory.free,memory.total,temperature.gpu --format=csv,noheader
    0 %, 0 %, 0 MiB, 11178 MiB, 11178 MiB, 45
    82 %, 28 %, 9099 MiB, 2079 MiB, 11178 MiB, 63
    '''
    output = fastfail_call(['nvidia-smi',
                            '--query-gpu=utilization.gpu,utilization.memory,memory.used,memory.free,memory.total,temperature.gpu',
                            '--format=csv,noheader,nounits'
                            ])
    if output is None:
        return []

    reader = csv.reader(io.StringIO(output), skipinitialspace=True)

    return [{
            'gpu_id': i,
            'gpu_perc': float(row[0]),
            'mem_perc': float(row[1]),
            'mem_used': int(row[2]),
            'mem_free': int(row[3]),
            'mem_tot': int(row[4]),
            'gpu_temperature': float(row[5]),
            } for i, row in enumerate(reader)]


def nvidia_smi_query_apps():
    '''
    nvidia-smi --query-compute-apps=pid,used_memory --format=csv,noheader,nounits
    15185, 9089

    nvidia-smi --query-compute-apps=pid,used_memory --format=csv,noheader
    15185, 9089 MiB
    '''
    output = fastfail_call(['nvidia-smi',
                            '--query-compute-apps=pid,used_memory',
                            '--format=csv,noheader,nounits'
                            ])
    if output is None:
        return []

    def _convert_row(row):
        pid = int(row[0])
        mem_used = int(row[1])

        p = psutil.Process(pid)
        app_name = p.environ().get('APP_NAME', '')
        proc_name = ' '.join(p.cmdline())
        create_time = int(p.create_time())
        running_time = get_running_time(p.create_time())

        return {
            'pid': pid,
            'app_name': app_name,
            'proc_name': proc_name,
            'mem_used': mem_used,
            'create_time': create_time,
            'running_time': running_time,
        }

    reader = csv.reader(io.StringIO(output), skipinitialspace=True)
    return [_convert_row(row) for row in reader]


def docker_stats():
    '''
    docker stats --no-stream --format {{.Name}},{{.CPUPerc}},{{.MemPerc}},{{.MemUsage}},{{.NetIO}},{{.BlockIO}},{{.PIDs}}
    DianAI,171.47%,5.43%,3.408GiB / 62.78GiB,62.5GB / 23.2GB,14.7GB / 3.82GB,/109
    MrHei2,0.00%,0.00%,716KiB / 62.78GiB,6.35GB / 164MB,20.3GB / 9.1GB,1
    '''
    output = fastfail_call(['docker',
                            'stats',
                            '--no-stream',
                            '--format',
                            "{{.Name}},{{.CPUPerc}},{{.MemPerc}},{{.MemUsage}},{{.NetIO}},{{.BlockIO}},{{.PIDs}}"
                            ])
    reader = csv.reader(io.StringIO(output), skipinitialspace=True)

    def _convert_row(row):
        mem_used, _, mem_limit = row[3].split()
        net_input, _, net_output = row[4].split()
        block_read, _, block_write = row[5].split()

        return {
            'name': row[0],
            'cpu_perc': float(row[1]),
            'mem_perc': float(row[2]),
            'mem_used': int(mem_used),
            'mem_limit': int(mem_limit),
            'net_input': int(net_input),
            'net_output': int(net_output),
            'block_read': int(block_read),
            'block_write': int(block_write),
        }

    return [_convert_row(row) for row in reader]


def docker_stats2(filters=None):
    client = docker.from_env()
    containers = client.containers.list()

    # Filter the containers by specified container names
    if isinstance(filters, list):
        containers = [c for c in containers if c.name in filters]

    def _convert_stats(c):
        stats = c.stats(stream=False)
        if c.name in g_pre_stats:
            cpu_stats = stats['cpu_stats']
            pre_cpu_stats = g_pre_stats[c.name]['cpu_stats']
            cpu_perc = (cpu_stats['cpu_usage']['total_usage'] - pre_cpu_stats['cpu_usage']
                        ['total_usage']) / (cpu_stats['system_cpu_usage'] - pre_cpu_stats['system_cpu_usage']) * 100.0
        else:
            cpu_perc = 0

        mem_used = stats['memory_stats']['stats']['rss']
        mem_limit = stats['memory_stats']['limit']

        net_input, net_output = 0, 0
        if 'networks' in stats:
            for _, v in stats['networks'].items():
                net_input += v['rx_bytes']
                net_output += v['tx_bytes']

        blk_read, blk_write = 0, 0
        for v in stats['blkio_stats']['io_service_bytes_recursive']:
            if v['op'] == 'Read':
                blk_read += v['value']
            elif v['op'] == 'Write':
                blk_write += v['value']

        g_pre_stats[c.name] = stats

        return {
            'name': c.name,
            'cpu_perc': float(cpu_perc),
            'mem_perc': float(mem_used) / float(mem_limit) * 100,
            'mem_used': int(mem_used),
            'mem_limit': int(mem_limit),
            'net_input': int(net_input),
            'net_output': int(net_output),
            'block_read': int(blk_read),
            'block_write': int(blk_write),
        }

    return [_convert_stats(c) for c in containers]


def disk_usage(filters=['/', '/disk']):
    mountpoints = filters
    devices = psutil.disk_partitions()
    mount_device = {}
    for m in mountpoints:
        for d in devices:
            if m == d.mountpoint:
                mount_device[m] = d.device

    def _disk_usage(m):
        try:
            du = psutil.disk_usage(m)
            return {
                'path': m,
                'device': mount_device[m],
                'total': du.total,
                'used': du.used,
                'free': du.free,
                'percent': du.percent,
            }
        except OSError as e:
            logger_runtime.critical(e)
            return {
                'path': m,
                'device': '',
                'total': 0,
                'used': 0,
                'free': 0,
                'percent': 0,
            }

    return [_disk_usage(m) for i, m in enumerate(mountpoints)]


def sys_load():
    vm = psutil.virtual_memory()

    return {
        'cpu_perc': psutil.cpu_percent(),
        'mem_perc': vm.percent,
        'mem_tot': vm.total,
        'mem_used': vm.used,
        'mem_free': vm.free,
        'mem_avail': vm.available,
    }


def get_container_stats(filters=None):
    apps = nvidia_smi_query_apps()
    container_stats = docker_stats2(filters=filters)

    for s in container_stats:
        s['pids'] = container_pids(s['name'])

    for app in apps:
        for s in container_stats:
            if app['pid'] in s['pids']:
                app['container'] = s['name']

    for s in container_stats:
        del s['pids']

    return (
        ('app', apps),
        ('containers', container_stats),
    )


TYPES_MAP = {
    'sysinfo': sys_static_info,
    'sysload': sys_load,
    'gpu': nvidia_smi_query_gpu,
    'disk': disk_usage,
    'containers': get_container_stats,
}


def parse_repeat(repeat):
    if repeat:
        r = repeat.split('/')
        n, i = int(r[0]), int(r[1])
    else:
        n, i = 1, None
    return n, i


def do_repeat(func, n, i):
    import time

    func()

    for x in range(n - 1):
        time.sleep(i)
        func()


def _do_query(type, stdout, filters=None):
    func = TYPES_MAP[type]
    data = func(filters=filters) if filters else func()

    def _handle_data(data, type):
        if stdout:
            print(json.dumps(data, ensure_ascii=False, indent=4))
            return

        # ${MESSAGE} = type
        # ${.SDATA.meta.*} = json data
        # logger_monitor.info(type, extra={'structured_data': {'meta': data}})

        # We prefer this one!
        # ${.SDATA.meta.type} = type
        # ${MESSAGE} = json data
        logger_monitor.info(json.dumps(data, ensure_ascii=False),
                            extra={'structured_data': {'meta': {'type': type}}})

    def _handle_one_or_more_data(data, type):
        if isinstance(data, list):
            for d in data:
                _handle_data(d, type)
        else:
            _handle_data(data, type)

    if isinstance(data, tuple):
        for t, d in data:
            _handle_one_or_more_data(d, t)
    else:
        _handle_one_or_more_data(data, type)


def do_query(type, stdout, repeat, filters=None):
    n, i = parse_repeat(repeat)

    def func(): return _do_query(type, stdout, filters=filters)
    do_repeat(func, n, i)


def _main(argv):
    prog = os.path.basename(sys.argv[0])
    parser = argparse.ArgumentParser(prog=prog)
    parser.add_argument('-v', '--version', action='version',
                        version='%(prog)s-' + __version__)
    subparsers = parser.add_subparsers(title='command', dest='command')

    parser_query = subparsers.add_parser(
        'query', help='Query system info, log them via syslog protocol')
    parser_query.add_argument('type', type=str, help='info type: %s' %
                              ', '.join(TYPES_MAP.keys()))
    parser_query.add_argument('-l', '--stdout', action='store_true', default=False,
                              help="Print pretty json to console instead of send a log")
    parser_query.add_argument('-r', '--repeat', type=str,
                              help='n/i repeat n times every i seconds')
    parser_query.add_argument('-f', '--filters', nargs='+', type=str,
                              help='Filter the disk paths for disk type; filter the container names for containers type')

    parser_ces = subparsers.add_parser(
        'create-esindex', help='Create elasticsearch index')
    parser_ces.add_argument('index', type=str, help='index name', nargs='?')
    parser_ces.add_argument('-l', '--esurl', help='Specify the elasticsearch url',
                            default=utils.ES_URL)
    parser_ces.add_argument(
        '-a', '--all', action='store_true', help='Create all')

    parser_des = subparsers.add_parser(
        'delete-esindex', help='Delete elasticsearch index')
    parser_des.add_argument('index', type=str, help='index name', nargs='?')
    parser_des.add_argument('-l', '--esurl', help='Specify the elasticsearch url',
                            default=utils.ES_URL)
    parser_des.add_argument(
        '-a', '--all', action='store_true', help='Create all')

    args = parser.parse_args(argv)

    if not args.command:
        parser.print_help()
    elif args.command == 'query':
        if args.type not in TYPES_MAP:
            parser_query.print_help()
        else:
            do_query(args.type, args.stdout, args.repeat, filters=args.filters)
    elif args.command == 'create-esindex':
        if args.all:
            utils.create_es_indicies(es_url=args.esurl)
        else:
            if args.index is None:
                parser_ces.print_help()
                return

            utils.create_es_index(args.index, es_url=args.esurl)
    elif args.command == 'delete-esindex':
        if args.all:
            utils.delete_es_indicies(es_url=args.esurl)
        else:
            if args.index is None:
                parser_des.print_help()
                return
            utils.delete_es_index(args.index, es_url=args.esurl)


def main():
    _main(sys.argv[1:])


if __name__ == '__main__':
    main()
